### DISCUSSION TITLE
Issue while setting up LIS with Noah-MP 4.0.1

### DISCUSSION THREAD

**Discussion by PrateekSharma99** [NASA-LIS/LISF#1266] [General]
Hi everyone I am trying to workout LIS with Noah-MP 4.0.1 but it stops and this error pops up:

![Screenshot (31)](https://user-images.githubusercontent.com/89891132/215012091-dd0b965b-1a88-4c9c-bea6-b2b43d51d21b.png)

I have attached the log file here along with the code that I am running. Can anyone help me out?
[https://drive.google.com/file/d/1fNmP54anOTAF60cESqfjI3er6BGyjps8/view?usp=sharing, https://drive.google.com/file/d/1iuajCxTrZKbF_pisbqKI81Ekf1aGOGW5/view?usp=sharing](url)

**Comment by dmocko** [comment]
Hi @Daduji123 

I have two suggestions of why this error is happening:

1) Do you mean to be doing forcing and other perturbations?  

   Since you are not running DA and only have one ensemble member per tile, I don't think you need to go this.  Try changing:

```
Forcing perturbation algorithm:                       "GMAO scheme"                           "GMAO scheme"                           "GMAO scheme"
State perturbation algorithm:                         "GMAO scheme"                           "GMAO scheme"                           "GMAO scheme"
Observation perturbation algorithm:                   "GMAO scheme"                           "GMAO scheme"                           "GMAO scheme"
```

to:

```
Forcing perturbation algorithm:                       none    none    none
State perturbation algorithm:                         none    none    none
Observation perturbation algorithm:                   none    none    none
```

2) The other issue might be that LIS is expecting forcing at a land point in your mask, but the MERRA2 forcing does not have data there.  Can you also please attach your ldt.config file?

**Comment by dmocko** [comment]
I also forgot to mention to look at the end of all your lislog.0??? files, for other error/warning messages.

**Reply by PrateekSharma99** [reply]
hi @dmocko the first suggestion that you gave worked out the model but the issue arising is that from the output attributes, soil moisture needs to be removed. If soil moisture is kept there, the model runs and stops after the first cycle and gives the following error. Any ideas on why does this happen?
![Screenshot (33)](https://user-images.githubusercontent.com/89891132/215713324-62c725b4-8935-4f3b-89f0-5349a5fb9ff7.png)
 I think the reason for this is the 4 layer soil moisture concept of lis it's not able to write outputs properly of soil moisture Further I tried changing the 

Output methodology:                        "2d gridspace"
Output naming style:                       "3 level hierarchy"

TO:::::

Output methodology:                        "1d tilespace
Output naming style:                       "WMO convention"
The model gave outputs of soil moisture in that case but its not clear as to why is this happening.

**Comment by dmocko** [comment]
Hi @Daduji123 

Can you please share your latest lis.config file as well as your MODEL_OUTPUT_TBL file?

**Reply by PrateekSharma99** [reply]
PFA
[Prateel.zip](https://github.com/NASA-LIS/LISF/files/10553393/Prateel.zip)

**Reply by PrateekSharma99** [reply]
Hi @dmocko I was able to solve the issue. I replaced this part in the model_output_tbl file that i have attached earlier with
SoilMoist:    1  m3/m3   -    2 0 0 4 201 1000    # Average layer soil moisture (kg/m2)

The model is now working fine but another issue is arising now. Earlier the HYMAP ie the routing part was turned off for the runs but now that the  model is up and running, I turned the part on but the model does not generate any outputs of the HYMAP routing. On a single cycle of 1 day the model writes the surface outputs but stops as soon as it starts writing the routing outputs. Iam not able to comprehend as to why this is happening. 
PFA the .config, log and the output table file that Iam using now.
[new.zip](https://github.com/NASA-LIS/LISF/files/10556962/new.zip)

**Reply by dmocko** [reply]
Yes, thanks for identifying the first issue.  Noah-MP-4.0.1 LSM code is currently written such that only units of m3/m3 are available for soil moisture output.  If you want kg/m2, some code will need to be added.  I can show you how to do that, if desired.

I'm not sure why you are crashing when writing the ROUTING output file.  The units in your MODEL OUTPUT TBL look good to me.

My only suggestion is to turn off all output in the MODEL OUTPUT TBL _other_ than the "Streamflow:".  Please try that, and let us know if you are getting output.  If you are, you can try to add each additional routing variable one by one to try to identify if there is a problem with a particular variable.

Can you also please share your ldt.config file?  From when you ran LDT.

**Comment by PrateekSharma99** [comment]
Thank you for the suggestions. I turned off all the output parameters other than streamflow and the model works fine. For all the others turned on separately the model gives out the same error and fails while writing the output of routing. I am not sure why this is happening and why the model is computing streamflow only. Any ideas as to why this is happening?

We are not using the ldt.config file but directly the LIS input file that I am not allowed to share. I apologize for that on my end.
I also tried including the reservoir part into the HYMAP2 with only the 'streamflow' turned on in  the output table I came up with a new issue regarding **HYMAP2 reservoir operation input time series size** 
How is it supposed to be defined is it supposed to be the number of lines in the reservoir file or simply 1?

**Reply by PrateekSharma99** [reply]
Also please show me how to change the units of soil moisture.

**Reply by PrateekSharma99** [reply]
#HYMAP2 router settings with reservoirs TURNED ON
Routing model:            "HYMAP2 router"

TEMPLATE model timestep:                                               "1hr" #Only used when LSM is not provided. HyMAP reads Runoff+Baseflow
HYMAP2 routing model time step:                                        "1hr"
HYMAP2 routing model output interval:                                  "1da"
HYMAP2 routing model restart interval:                                 "1mo"
HYMAP2 routing method:                                                 "local inertia" # "kinematic", "diffusive", or "local inertia"
HYMAP2 routing model time step method:                                 "adaptive" # "constant" or "adaptive"
HYMAP2 routing model adaptive time step alfa coefficient:              1.0
HYMAP2 run in ensemble mode:                                            0
HYMAP2 routing model linear reservoir flag:                             1
HYMAP2 routing model evaporation option:                               none # "penman"
HYMAP2 routing model restart file:                                    ./output/LIS_MERRA2/ROUTING/200012/LIS_RST_HYMAP2_router_200012312300.d01.nc
HYMAP2 routing model start mode:                                       coldstart 

HYMAP2 floodplain dynamics:                                            1  # enter 1 - compute floodplain dynamics; or 0 - do not compute floodplain dynamics
HYMAP2 reservoir operation option:                                     1 # linear reservoir flag: enter 1 - use; or 0 - do not use linear reservoirs
HYMAP2 number of reservoirs:                                           1
HYMAP2 reservoir operation input time series size:                     5
HYMAP2 reservoir operation input directory:                           ./reservoirdata/
HYMAP2 reservoir operation header filename:                           ./reservoirdata/header_test.txt
HYMAP2 reservoir operation input data type:                           "water level" # "streamflow" #
HYMAP2 enable 2-way coupling: 0
HYMAP2 routing model dwi flag: 0


Since the model is working when streamflow is the only output, I tried including the reservoirs into it wherin I am encountred some issues. I 
tried working out 1 single reservoir for 5 days (PFA the header file and reservoir file in the mail attached ) but the model does not run 
and stops with the following error:

 [ERR] HYMAP2 reservoir operation input time series length: not defined
  Stopping.
 [ERR] endrun is being called

I tried changing the time series size option to 1 but still there was the same error. I looked at the documentation of LISF-7.4 wherein its mentioned that
the HYMAP2 reservoir operation input time series size: specifies the time series length (or number of time steps in the inputs files) when using reservoir 
operations with the HyMAP2 router. Based on this there should be no error and reservoirs should work just fine. What would be correct way to define the size
of time series or is it so that my code is missing something?

**Reply by dmocko** [reply]
To add units of "kg m-2" of soil moisture to the output, edit this file:  `LISF/lis/surfacemodels/land/noahmp.4.0.1/NoahMP401_main.F90`

Add the line with `unit="kg m-2"` below (existing lines are shown for context on where this line should go):

```
            ![ 7] output variable: smc (unit=m3/m3). ***  volumetric soil moisture
            do i=1, NOAHMP401_struc(n)%nsoil
                call LIS_diagnoseSurfaceOutputVar(n, t, LIS_MOC_SOILMOIST, value = NOAHMP401_struc(n)%noahmp401(t)%smc(i), &
                                                  vlevel=i, unit="m^3 m-3", direction="-", surface_type = LIS_rc%lsm_index)
                call LIS_diagnoseSurfaceOutputVar(n, t, LIS_MOC_SOILMOIST, value = NOAHMP401_struc(n)%noahmp401(t)%smc(i)*tmp_sldpth(i)*LIS_CONST_RHOFW, &
                                                  vlevel=i, unit="kg m-2", direction="-", surface_type = LIS_rc%lsm_index)
            end do

```
Re-compile LIS, and then change the units for `SoilMoist:` in your MODEL OUTPUT TBL file back to "kg/m2".

**Comment by PrateekSharma99** [comment]
Hi @dmocko Thanks for the suggestions. I was able to solve the error in routing part by changing the model output parameters to the following new settings:
![image](https://user-images.githubusercontent.com/89891132/216530194-bb552336-0c91-495c-b03f-1b7857315a54.png)
Without the reservoir flag turned on, the model now works completely fine. Once again thankyou for helping me out with the issues.
